test/test_dtype.py

```diff
@@ -3,28 +3,66 @@
 from tinygrad.helpers import getenv, DType, DEBUG
 from tinygrad.lazy import Device
 from tinygrad.tensor import Tensor, dtypes
+from typing import List, Optional
 from extra.utils import OSX, temp
+import copy
 
 def _test_to_np(a:Tensor, np_dtype, target):
-  print(a)
+  if DEBUG >= 2: print(a)
   na = a.numpy()
-  print(na, na.dtype, a.lazydata.realized)
-  assert na.dtype == np_dtype
-  np.testing.assert_allclose(na, target)
+  if DEBUG >= 2: print(na, na.dtype, a.lazydata.realized)
+  try:
+    assert na.dtype == np_dtype
+    np.testing.assert_allclose(na, target)
+  except AssertionError as e:
+    raise AssertionError(
+      f"\ntensor {a.numpy()} does not match target {target} with np_dtype {np_dtype}"
+    ) from e
+
+def _assert_eq(tensor:Tensor, target_dtype:DType, target):
+  if DEBUG >= 2: print(tensor.numpy())
+  try:
+    assert tensor.dtype == target_dtype
+    np.testing.assert_allclose(tensor.numpy(), target)
+  except AssertionError as e:
+    raise AssertionError(
+      f"\ntensor {tensor.numpy()} dtype {tensor.dtype} does not match target {target} with dtype {target_dtype}"
+    ) from e
 
 def _test_op(fxn, target_dtype:DType, target):
-  c = fxn()
-  if DEBUG >= 2: print(c.numpy())
-  assert c.dtype == target_dtype
-  np.testing.assert_allclose(c.numpy(), target)
+  _assert_eq(fxn(), target_dtype, target)
+
+def _test_cast(a:Tensor, target_dtype:DType, target):
+  _test_op(lambda: a.cast(target_dtype), target_dtype, target)
+# tests no-op casts from source_dtype to target_dtypes
+def _test_casts_from(tensor_contents:List, source_dtype:DType, target_dtypes:List[DType], target_contents:Optional[List]=None):
+  if target_contents is None: target_contents = copy.deepcopy(tensor_contents)
+  list(map(
+    lambda t_dtype: _test_cast(Tensor(tensor_contents, dtype=source_dtype), t_dtype, target_contents),
+    target_dtypes
+  ))
+# tests no-op casts from source_dtypes to target_dtype
+def _test_casts_to(tensor_contents:List, source_dtypes:List[DType], target_dtype:DType, target_contents:Optional[List]=None):
+  if target_contents is None: target_contents = copy.deepcopy(tensor_contents)
+  list(map(
+    lambda s_dtype: _test_cast(Tensor(tensor_contents, dtype=s_dtype), target_dtype, target_contents),
+    source_dtypes
+  ))
+
+def _test_ops(a_dtype:DType, b_dtype:DType, target_dtype:DType):
+  _assert_eq(
+    Tensor([1,2,3,4], dtype=a_dtype) + Tensor([1,2,3,4], dtype=b_dtype),
+    target_dtype, [2,4,6,8]
+  )
+  _assert_eq(
+    Tensor([1,2,3,4], dtype=a_dtype) * Tensor([1,2,3,4], dtype=b_dtype),
+    target_dtype, [1,4,9,16]
+  )
+  _assert_eq(
+    Tensor([[1,2],[3,4]], dtype=a_dtype) @ Tensor.eye(2, dtype=b_dtype),
+    target_dtype, [[1,2],[3,4]]
+  )
```

Comment:
Can you just one line these?

Reply:
Yep, no problem. One-lined these and a few other things that were easy to read.